File: built-in plugins dsp/mb-sampler-guitar.cpp
/**
 * MB Guitar Sampler
 * Category : instrument
 * Type     : sampler
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Acoustic and electric guitars
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SAMPLER_GUITAR_H
#define MB_SAMPLER_GUITAR_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSamplerGuitar : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-sampler-guitar";
    static constexpr const char* PLUGIN_NAME    = "MB Guitar Sampler";
    static constexpr const char* PLUGIN_TYPE    = "sampler";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float type = 0.5f;  // range [0, 1]
    float pick = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSamplerGuitar() = default;
    ~MbSamplerGuitar() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.type = std::clamp(params.type, 0f, 1f);
        params.pick = std::clamp(params.pick, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Guitar Sampler
        return input;
    }
};

#endif // MB_SAMPLER_GUITAR_H

File: built-in plugins dsp/mb-sampler-guitar.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSamplerGuitarPlugin: PluginDefinition = { id: 'mb-sampler-guitar', slug: 'mb-sampler-guitar', name: 'MB Guitar Sampler', category: 'instrument', type: 'sampler', version: '1.0.0', description: 'Acoustic and electric guitars', author: 'Max Booster', oscillators: [], envelope: { attack: 0.005, decay: 0.5, sustain: 0.5, release: 0.4 }, parameters: [{ id: 'type', name: 'Type', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: false }, { id: 'pick', name: 'Pick Position', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { type: 0.5, pick: 0.5, volume: 0.8 } };

export default MbSamplerGuitarPlugin;
